libsplinter/src/service/dispatch/task_job_executor.rs

```diff
@@ -0,0 +1,138 @@
+    /// Construct a new builder.
+    pub fn new() -> Self {
+        Self {
+            thread_pool_builder: ThreadPoolBuilder::new(),
+            prefix: None,
+        }
+    }
+
+    /// Set the prefix for the worker threads in the pool.
+    pub fn with_prefix(mut self, prefix: String) -> Self {
+        self.prefix = Some(prefix);
+        self
+    }
+
+    /// Set the size of the pool.
+    pub fn with_size(mut self, size: usize) -> Self {
+        self.thread_pool_builder = self.thread_pool_builder.with_size(size);
+        self
+    }
+
+    /// Constructs the task pool.
+    ///
+    /// # Errors
+    ///
+    /// Will return an [`InvalidStateError`] if the pool has not be configured with a size.
+    pub fn build(self) -> Result<MessageHandlerTaskPool, InvalidStateError> {
+        let thread_pool = self
+            .thread_pool_builder
+            .with_prefix(
+                self.prefix
+                    .unwrap_or_else(|| "MessageHandlerTaskPool".to_string()),
+            )
+            .build()
+            .map_err(|err| InvalidStateError::with_message(err.to_string()))?;
+
+        Ok(MessageHandlerTaskPool { thread_pool })
+    }
+}
+
+/// A pool of [`MessageHandlerTaskRunner`] instances.
+pub struct MessageHandlerTaskPool {
+    thread_pool: ThreadPool,
+}
+
+impl MessageHandlerTaskPool {
+    /// Returns a [`MessageHandlerTaskRunner`] instance.
+    pub fn task_runner(&self) -> impl MessageHandlerTaskRunner + Send {
+        JobExecutorMessageHandlerTaskRunner::new(self.thread_pool.executor())
+    }
+}
+
+impl ShutdownHandle for MessageHandlerTaskPool {
+    fn signal_shutdown(&mut self) {
+        self.thread_pool.shutdown_signaler().shutdown();
+    }
+
+    fn wait_for_shutdown(self) -> Result<(), InternalError> {
+        self.thread_pool.join_all();
+
+        Ok(())
+    }
+}
+
+struct JobExecutorMessageHandlerTaskRunner {
+    job_executor: JobExecutor,
+}
+
+impl JobExecutorMessageHandlerTaskRunner {
+    fn new(job_executor: JobExecutor) -> Self {
+        Self { job_executor }
+    }
+}
+
+impl MessageHandlerTaskRunner for JobExecutorMessageHandlerTaskRunner {
+    fn execute(
+        &self,
+        message_handler_factory: &dyn MessageHandlerFactory<
+            MessageHandler = Box<dyn MessageHandler<Message = Vec<u8>>>,
+        >,
+        sender_factory: &dyn MessageSenderFactory<Vec<u8>>,
+        to_service: FullyQualifiedServiceId,
+        from_service: FullyQualifiedServiceId,
+        message: Vec<u8>,
+    ) -> Result<(), InternalError> {
+        let factory = message_handler_factory.clone_boxed();
+        let sender_factory = sender_factory.clone_boxed();
+
+        self.job_executor.execute(move || {
+            let mut handler = factory.new_handler();
+            let sender = match sender_factory.new_message_sender(&to_service) {
+                Ok(sender) => sender,
+                Err(err) => {
+                    error!("Unable to create new message sender: {}", err);
```

Comment:
We lose the information about the message being sent; the message itself is just bytes but we could log from_service/to_service. Same with the error!() below.

Reply:
Updated